test/0dt/mzcompose.py

```diff
@@ -865,6 +863,343 @@ def workflow_basic(c: Composition) -> None:
+        GRANT ALL ON CLUSTER cluster TO materialize;
+        ALTER SYSTEM SET cluster = cluster;
+    """,
+        service="mz_old",
+        port=6877,
+        user="mz_system",
+    )
+
+    start_time = time.time()
+    c.testdrive(
+        dedent(
+            f"""
+        > SET CLUSTER = cluster;
+
+        > CREATE CONNECTION IF NOT EXISTS kafka_conn FOR KAFKA BROKER '${{testdrive.kafka-addr}}', SECURITY PROTOCOL = 'PLAINTEXT';
+        > CREATE CONNECTION IF NOT EXISTS csr_conn FOR CONFLUENT SCHEMA REGISTRY URL '${{testdrive.schema-registry-url}}';
+
+        $ kafka-create-topic topic=kafka-large
+        $ kafka-ingest format=bytes key-format=bytes key-terminator=: topic=kafka-large repeat={count}
+        key0A,key${{kafka-ingest.iteration}}:value0A,${{kafka-ingest.iteration}}
+        > CREATE SOURCE kafka_source
+          IN CLUSTER cluster
+          FROM KAFKA CONNECTION kafka_conn (TOPIC 'testdrive-kafka-large-${{testdrive.seed}}');
+
+        > CREATE TABLE kafka_source_tbl (key1, key2, value1, value2)
+          FROM SOURCE kafka_source (REFERENCE "testdrive-kafka-large-${{testdrive.seed}}")
+          KEY FORMAT CSV WITH 2 COLUMNS DELIMITED BY ','
+          VALUE FORMAT CSV WITH 2 COLUMNS DELIMITED BY ','
+          ENVELOPE UPSERT;
+        # TODO: This should also work with a default index
+        # > CREATE DEFAULT INDEX ON kafka_source_tbl
+        > SELECT COUNT(*) FROM kafka_source_tbl
+        {count}
+        """
+        )
+    )
+    for i in range(1, repeats):
+        c.testdrive(
+            dedent(
+                f"""
+        $ kafka-ingest format=bytes key-format=bytes key-terminator=: topic=kafka-large repeat={count}
+        key{i}A,key{i}${{kafka-ingest.iteration}}:value{i}A,${{kafka-ingest.iteration}}
+        > SELECT COUNT(*) FROM kafka_source_tbl
+        {count*(i+1)}
+        """
+            )
+        )
+
+    elapsed = time.time() - start_time
+    print(f"initial ingestion took {elapsed} seconds")
+
+    with c.override(
+        Materialized(
+            name="mz_new",
+            sanity_restart=False,
+            deploy_generation=1,
+            system_parameter_defaults=SYSTEM_PARAMETER_DEFAULTS,
+            restart="on-failure",
+            external_metadata_store=True,
+            # environment_extra=["FAILPOINTS=hydration=sleep(12000)"],
+        )
+    ):
+        c.up("mz_new")
+        start_time = time.time()
+        c.await_mz_deployment_status(DeploymentStatus.READY_TO_PROMOTE, "mz_new")
+        elapsed = time.time() - start_time
+        print(f"re-hydration took {elapsed} seconds")
+        c.promote_mz("mz_new")
+        start_time = time.time()
+        c.await_mz_deployment_status(
+            DeploymentStatus.IS_LEADER, "mz_new", sleep_time=None
+        )
+        elapsed = time.time() - start_time
+        print(f"promotion took {elapsed} seconds")
+        print("sleeping so that the global query timestamp can advance")
+        time.sleep(5)
```

Comment:
This seems weird to me. How necessary is this for the test to be reliable? I feel it's a much stronger test without this sleep, as it keeps us honest about whether the "zero" in "zero downtime" is really close to zero (rather than 5s).

Reply:
This sleep was suggested by Aljoscha, it's what made the test work at all: https://materializeinc.slack.com/archives/C071VTZS1M5/p1732197313421659?thread_ts=1731578710.226209&cid=C071VTZS1M5
I feel like this is extremely brittle too.

Reply:
Removed the sleep, with a proper index it still seems to work. Without the index+view it did not.

Reply:
Ok cool! Something to look into for sure, but I feel good about the test as written—it's definitely testing the right thing.